packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs

```diff
@@ -3,118 +3,163 @@ use error_stack::{IntoReport, Report, Result, ResultExt};
-                SELECT link_type_version_id, target_entity_id AS single_link, NULL AS multi_link FROM single_links 
-                UNION 
-                SELECT link_type_version_id, NULL AS single_link, target_entity_ids AS multi_link from multi_links
-                "#,
-            &[&source_entity_id, link_type_uri.base_uri(), &version],
+            -- Gather all single-links
+            WITH single_links AS (
+                SELECT link_type_version_id, target_entity_id
+                FROM links
+                INNER JOIN ids ON ids.version_id = links.link_type_version_id
+                WHERE active AND NOT multi AND source_entity_id = $1 AND base_uri = $2 AND "version" = $3
+            ),
+            -- Gather all multi-links
+            multi_links AS (
+                SELECT link_type_version_id, ARRAY_AGG(target_entity_id ORDER BY multi_order ASC) AS target_entity_ids
+                FROM links
+                INNER JOIN ids ON ids.version_id = links.link_type_version_id
+                WHERE active AND multi AND source_entity_id = $1 AND base_uri = $2 AND "version" = $3 GROUP BY link_type_version_id
+            )
+            -- Combine single and multi links with null values in rows where the other doesn't exist
+            SELECT link_type_version_id, target_entity_id AS single_link, NULL AS multi_link
+            FROM single_links
+            UNION
+            SELECT link_type_version_id, NULL AS single_link, target_entity_ids AS multi_link
+            FROM multi_links
+            "#,
+            &[&source_entity_id, link_type_uri, &i64::from(link_type_version)],
         )
         .await
         .into_report()
         .change_context(QueryError)
         .attach_printable(source_entity_id)
         .attach_printable(link_type_uri.clone())?;
 
-        let val: (Option<EntityId>, Option<Vec<EntityId>>) = (link.get(1), link.get(2));
-        match val {
-            (Some(entity_id), None) => Ok(Outgoing::Single(entity_id)),
-            (None, Some(entity_ids)) => Ok(Outgoing::Multiple(entity_ids)),
-            _ => Err(Report::new(QueryError)
+    let val: (Option<EntityId>, Option<Vec<EntityId>>) = (link.get(1), link.get(2));
+    let outgoing = match val {
+        (Some(entity_id), None) => Outgoing::Single(entity_id),
+        (None, Some(entity_ids)) => Outgoing::Multiple(entity_ids),
+        _ => {
+            return Err(Report::new(QueryError)
                 .attach_printable(source_entity_id)
-                .attach_printable(link_type_uri.clone())),
+                .attach_printable(link_type_uri.clone()));
+        }
+    };
+    Ok(vec![Links::new(
+        [(
+            VersionedUri::new(link_type_uri.to_string(), link_type_version),
+            outgoing,
+        )]
+        .into(),
+    )])
+}
+
+// TODO: we should probably support taking PersistedEntityIdentifier here as well as an EntityId
```

Comment:
Not entirely sure what this comment refers to? 
Doe sit mean that in place of having the `LinkQuery` specify a `source_entity_id` which is of type `EntityId`, we can tak in `PersistedEntityIdentifier`?

Reply:
Yes 🙂